### PR TITLE
[Activation] Fix TxMempoolBlockDoublespend

### PR DESCRIPTION
**Pull request by dangershony**

### CODE REVIEW

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 1]
```diff
@@ -386,10 +386,10 @@ private async Task<ConnectNewHeadersResult> PresentHeadersLockedAsync(List<Block
                 // Resync in case can't connect.
                 await this.ResyncAsync().ConfigureAwait(false);
             }
-            catch (ConsensusErrorException exception)
+            catch (HeaderInvalidException)
```
Comment (majikandy) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: I find the mismatch between ConsensusErrors.BannedHash.Throw and HeaderInvalidException awkward to follow. I assume the .Throw throws that exception but I would say that it isn’t obvious.

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 2]
```diff
@@ -386,10 +386,10 @@ private async Task<ConnectNewHeadersResult> PresentHeadersLockedAsync(List<Block
...
             {
-                this.logger.LogDebug("Peer's header is invalid. Peer will be banned and disconnected. Error: {0}.", exception.ConsensusError);
-                this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, this.connectionManager.ConnectionSettings.BanTimeSeconds, $"Peer presented invalid header, error: {exception.ConsensusError}.");
+                this.logger.LogDebug("Peer's header is invalid. Peer will be banned and disconnected.");
+                this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, this.connectionManager.ConnectionSettings.BanTimeSeconds, $"Peer presented invalid header.");
```
Comment (majikandy) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: Magic strings, potentially should be constants

Comment (majikandy) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: And no need for the $ string interpolation

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: They sort of have the same meaning so `HeaderInvalidException` is more general.

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 3]
```diff
@@ -1097,7 +1097,7 @@ private ChainedHeader CreateAndValidateNewChainedHeader(BlockHeader currentBlock
             if (result.Error != null)
             {
                 this.logger.LogTrace("(-)[INVALID_HEADER]");
-                result.Error.Throw();
+                throw new ConsensusRuleException(result.Error);
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Shouldn't this be `HeaderInvalidException` as we validated the header above?

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs [hunk 4]
```diff
@@ -1087,7 +1087,7 @@ private ChainedHeader CreateAndValidateNewChainedHeader(BlockHeader currentBlock
             if (this.invalidHashesStore.IsInvalid(newHeaderHash))
             {
                 this.logger.LogTrace("(-)[HEADER_HASH_MARKED_INVALID]");
-                ConsensusErrors.BannedHash.Throw();
+                throw new HeaderInvalidException();
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: Perhaps pass the "HEADER_HASH_MARKED_INVALID" message to the exception?

File: src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs [hunk 5]
```diff
@@ -28,6 +28,23 @@ public ConnectHeaderException() : base()
         }
     }
 
+    public class HeaderInvalidException : ConsensusException
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs]: Document these new exceptions.

File: src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs [hunk 6]
```diff
@@ -28,6 +28,23 @@ public ConnectHeaderException() : base()
...
+    {
+        public HeaderInvalidException() : base()
+        {
+        }
+    }
+
+    public class ConsensusRuleException : ConsensusException
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs]: Document these new exceptions.

File: src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs [hunk 7]
```diff
@@ -386,11 +386,16 @@ private async Task<ConnectNewHeadersResult> PresentHeadersLockedAsync(List<Block
                 // Resync in case can't connect.
                 await this.ResyncAsync().ConfigureAwait(false);
             }
-            catch (ConsensusErrorException exception)
+            catch (ConsensusRuleException exception)
             {
                 this.logger.LogDebug("Peer's header is invalid. Peer will be banned and disconnected. Error: {0}.", exception.ConsensusError);
                 this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, this.connectionManager.ConnectionSettings.BanTimeSeconds, $"Peer presented invalid header, error: {exception.ConsensusError}.");
             }
+            catch (HeaderInvalidException)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs]: If we are going to pass the specific message to the exception we might want to write this to the log as well?

File: src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs (hunk 4 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs]: No this is a specific exception.

**Comment by noescape00** [comment]
looks good. 

however I'm against spawning more and more different types of consensus errors.

File: src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs [hunk 8]
```diff
@@ -28,6 +28,30 @@ public ConnectHeaderException() : base()
...
+    {
+        public HeaderInvalidException() : base()
+        {
+        }
+    }
+
+    /// <summary>
+    /// An exception that is contains exception coming from the <see cref="IConsensusRuleEngine"/> execution.
+    /// </summary>
+    public class ConsensusRuleException : ConsensusException
```
Comment (noescape00) [inline src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs]: this is basically a copy-paste of ConsensusErrorException

Comment (dangershony) [reply src/Stratis.Bitcoin/Consensus/ConsensusExceptions.cs]: LOL yes it is even the name is wrong they should be swapped but didnt want to make such a big change yet.

**Pull request merged**